Add support for AIDL binder interface

mlehtima opened the pull request.

spiiroin commented on core/hybrisadaptor.cpp:
```diff
@@ -65,6 +69,11 @@ const char* const sensors_2_callback_ifaces[] = {
     NULL
 };
 
+const char* const sensors_aidl_callback_ifaces[] = {
```
qt nit: "char* foo" -> "char *foo"

mlehtima replied:
Fixed

spiiroin commented on core/hybrisadaptor.cpp:
```diff
@@ -692,7 +704,21 @@ GBinderLocalReply *HybrisManager::sensorCallbackHandler(
     qCInfo(lcSensorFw) << "sensorCallbackHandler";
     const char *iface = gbinder_remote_request_interface(req);
-    if (iface && (!strcmp(iface, SENSOR_BINDER_SERVICE_IFACE_2_0) ||
+    if (iface && (!strcmp(iface, SENSOR_BINDER_SERVICE_IFACE_AIDL))) {
```
nit: superfluous () around strcmp()

mlehtima replied:
Fixed

spiiroin commented on core/hybrisadaptor.cpp:
```diff
@@ -692,7 +704,21 @@ GBinderLocalReply *HybrisManager::sensorCallbackHandler(
+        }
+        *status = GBINDER_STATUS_OK;
+        qCInfo(lcSensorFw) << "sensorCallbackHandler valid sensor interface";
+    } else if (iface && (!strcmp(iface, SENSOR_BINDER_SERVICE_IFACE_2_0) ||
```
qt nit: '||' goes to next line

mlehtima replied:
Fixed

spiiroin commented on core/hybrisadaptor.cpp:
```diff
@@ -789,11 +850,16 @@ void HybrisManager::binderDied(GBinderRemoteObject *, void *user_data)
+        m_serviceManagerAidl = gbinder_servicemanager_new(SENSOR_BINDER_SERVICE_DEVICE_AIDL);
+    }
+
+    if (gbinder_servicemanager_wait(m_serviceManagerHidl, -1) &&
```
qt nit: '&&' goes to next line

mlehtima replied:
Fixed

mlehtima committed:
[hybrisadaptor] Split hybris backends to separate classes. JB#61406

Don't wait for hwbinder servicemanager when checking if hidl backend is
supported. On new devices hwbinder servicemanager is not present anymore
and waiting for servicemanager would cause the service to wait forever.

mlehtima committed:
[hybrisadaptor] Add support for AIDL binder interface. JB#61406

mlehtima commented:
Reworked the PR to split the backends into separate classes to make the code more readable.

monich commented on core/hybrisbackend.h:
```diff
@@ -0,0 +1,75 @@
+const size_t maxEvents = 64;
+}
+
+#ifdef USE_BINDER
```
do these implementation specific headers need to be part of the interface?

mlehtima replied:
There would be a lot more refactoring needed to get rid of these since HybrisAdaptor uses sensors_event_t which comes from these headers.

monich commented on core/hybrisbackend.h:
```diff
@@ -0,0 +1,75 @@
+    HybrisBackend(HybrisManager *manager, QObject *parent = 0);
+    virtual void initialize() = 0;
+    virtual void cleanup() = 0;
+    int sensorCount();
```
looks a bit strange/inconsistent that this method isn't virtual like everything else🤔

mlehtima replied:
Changed to virtual.

abranson commented on core/hybrisadaptor.cpp:
```diff
@@ -629,43 +410,8 @@ void HybrisManager::cleanup()
-    }
-#endif
-
     delete[] m_sensorState;
```
Is this stuff still done when the binderDieds happen in the backends? It seems like the rest of this is copied there, but this stuff isn't done anymore. Is that an issue?

mlehtima replied:
Changed the code to terminate sensorfw in case binder dies which improves recovering sensor connection to lipstick.

abranson commented on core/hybrisbackend_binder.cpp:
```diff
@@ -0,0 +1,171 @@
+        g_free((void*)m_sensorArray[i].requiredPermission.data.str);
+    }
+    delete[] m_sensorArray;
+    m_sensorArray = NULL;
```
Should m_sensorCount should be set to 0 here?

pvuorela replied:
Yea that would be symmetric. Now it's initialized from initialize(), cleanup() could then deinit. Though the init is in derived classes which makes this spread out on multiple files. Got myself confused for a moment as nothing in this class assigns to m_sensorCount but reads the value.

Could be slightly simpler to follow with some actual container type rather than plain array.

mlehtima replied:
Done

pvuorela commented on core/hybrisbackend_hal.h:
```diff
@@ -0,0 +1,63 @@
+****************************************************************************/
+
+#ifndef HybrisBackendHal_H
+#define HybrisBackendHal_H
```
Minor detail but the CamelCased if guard is slightly unusual. Suspect it might have been some search & replace thing here. Some other ones also in the PR.

mlehtima replied:
Fixed

pvuorela commented on core/hybrisbackend_hal.h:
```diff
@@ -0,0 +1,63 @@
+public:
+    HybrisBackendHal(HybrisManager *manager, QObject *parent = 0);
+    ~HybrisBackendHal();
+    void initialize();
```
Bonus points for marking override on the reimplemented methods.

mlehtima replied:
Done

pvuorela commented on core/hybrisbackend.h:
```diff
@@ -0,0 +1,75 @@
+    virtual float resolution(int index) = 0;
+    virtual int setActive(int handle, bool active) = 0;
+    virtual int setDelay(int handle, int64_t delay_ns) = 0;
+    virtual void eventReaderThreadImpl() = 0;
```
I can underdstand Impl in e.g. class names in comparison to abstract ones, but on one method it looks strange. All the methods are "Impl" in the subclasses?

The name neither tells what it does and not sure even the thread part.

mlehtima replied:
Renamed the method to readEvents.

pvuorela commented on core/hybrisbackend_binder.cpp:
```diff
@@ -0,0 +1,171 @@
+**
+****************************************************************************/
+
+#ifdef USE_BINDER
```
Should anything_binder.cpp be even ever compiled without USE_BINDER?

mlehtima replied:
Reworked this so only used files are included in hybris.pro.

pvuorela commented on core/hybrisbackend_binder.cpp:
```diff
@@ -0,0 +1,171 @@
+    HybrisBackend *backend = NULL;
+    for (int i = 0; i < 5; ++i) {
+        if (HybrisBackendBinderAidl::isSupported()) {
+            backend = qobject_cast<HybrisBackend *>(new HybrisBackendBinderAidl(manager));
```
The backends should all inherit HybrisBackend, thus not need any casting? Applies to all getBackend().

mlehtima replied:
Fixed

pvuorela commented on core/hybrisbackend_binder.cpp:
```diff
@@ -0,0 +1,171 @@
+
+HybrisBackend *getBackend(HybrisManager *manager)
+{
+    HybrisBackend *backend = NULL;
```
Bonus points for nullptr.

mlehtima replied:
Replaced all NULL with nullptr in the files I modified.

pvuorela reviewed (commented):
Some quick comments, didn't have time to go through all yet.

pvuorela commented on core/hybrisadaptor.cpp:
```diff
@@ -564,39 +370,14 @@ void HybrisManager::cleanup()
-        gbinder_remote_reply_unref(reply);
-        gbinder_local_request_unref(req);
+    if (m_backend) {
+        m_backend->cleanup();
```
This seems to be the only external use for the backend cleanup(). Since it's followed by deletion, could just make the deletion do cleanup and have one less method in public api? Internally the backends can keep such if they need to clean up state in the middle.

Then again now it's symmetric with separate initialize() but that also seems to be called right after getBackend() so the ctors could just do that?

mlehtima replied:
Removed the public cleanup() from HybrisBackend.

pvuorela commented on core/hybrisbackend.h:
```diff
@@ -0,0 +1,75 @@
+    int m_sensorCount;
+};
+
+HybrisBackend *getBackend(HybrisManager *manager);
```
Could this be a static method in HybrisBackend? Seeing the call in HybrisManager I first thought it was a method in that class and was confused for a while where it's defined.

mlehtima replied:
The method is now static member of HybrisBackend.

pvuorela commented on core/hybrisbackend.h:
```diff
@@ -0,0 +1,75 @@
+    HybrisBackend(HybrisManager *manager, QObject *parent = 0);
+    virtual void initialize() = 0;
+    virtual void cleanup() = 0;
+    virtual int sensorCount();
```
A single non-abstract virtual method is slightly peculiar. Does the trick with protected m_sensorCount but leaves me stll wondering would it be simpler if the backends just implemented keeping the count themselves. Or alternatively should this be non-virtual.

No biggie though.

mlehtima replied:
Kept this as is to remove need to redefine the method again in all backends..

pvuorela commented on core/hybrisbackend.h:
```diff
@@ -0,0 +1,75 @@
+    virtual int minDelay(int index) = 0;
+    virtual float maxRange(int index) = 0;
+    virtual float resolution(int index) = 0;
+    virtual int setActive(int handle, bool active) = 0;
```
The api with almost everything fetched as index might be cleaner with some sensor type returned by index and that having this sort of methods.

For another detail almost all is having index parameter but then couple things are based on handles. Then internally the code does some indexForType() -> handle() resolving. Didn't go through all but just wondering if there needs to be a handle exposed. Index is sort of handle too.

This more as suggestion rather than requirement to change.

mlehtima replied:
Didn't change the API now because I want to get the changes in.

pvuorela commented on core/hybrisbackend_binder.cpp:
```diff
@@ -0,0 +1,171 @@
+    int index = m_manager->indexForHandle(eve->sensor);
+    const struct sensor_t *sensor = &m_sensorArray[index];
+    if (sensor->flags & SENSOR_FLAG_WAKE_UP) {
+        return true;
```
Minor thing but this is classic 'if (isTrue()) return true; else return false;' 

Could be just  "return (sensor->flags & SENSOR_FLAG_WAKE_UP);"

mlehtima replied:
Changed.

pvuorela commented on core/hybrisbackend_binder.h:
```diff
@@ -0,0 +1,69 @@
+    float maxRange(int index);
+    float resolution(int index);
+    virtual int setActive(int handle, bool active) = 0;
+    virtual int setDelay(int handle, int64_t delay_ns) = 0;
```
No point in redeclaring abstract methods in parent class as still abstract? Same for initialize() and needsReaderThread().

mlehtima replied:
Removed the unnecessary redeclarations.

pvuorela commented on core/hybrisbackend_binder.h:
```diff
@@ -0,0 +1,69 @@
+    Q_OBJECT
+public:
+    HybrisBackendBinder(HybrisManager *manager, QObject *parent = 0);
+    virtual ~HybrisBackendBinder() = 0;
```
This exists? And dtors as = 0 in general would be strange.

mlehtima replied:
Replaced with empty virtual destructors to prevent complains by compiler.

pvuorela commented on core/hybrisbackend.h:
```diff
@@ -0,0 +1,75 @@
+
+class HybrisManager;
+
+class HybrisBackend : public QObject
```
Is there actually any use for the QObject inheritance? Not spotting signals nor slots, and the instance is explicitly deleted, not needing parenting.

mlehtima replied:
Removed QObject inheritance.

pvuorela reviewed (commented):
Bunch of more comments.

mlehtima committed:
[hybrisadaptor] Restart sensorfw if binder service dies. JB#61406

mlehtima committed:
[hybrisadaptor] Improve handling binder error cases. JB#61406
